code/controllers/subsystem.dm

```diff
@@ -254,17 +254,28 @@
 /// Called after the config has been loaded or reloaded.
 /datum/controller/subsystem/proc/OnConfigLoad()
 
-//used to initialize the subsystem AFTER the map has loaded
-/datum/controller/subsystem/Initialize()
+/**
+ * Used to initialize the subsystem.
+ * failed - Optional: If true, tell everyone that the subsystem failed to init and cancel firing.
+ */
+/datum/controller/subsystem/Initialize(failed = FALSE)
 	initialized = TRUE
-	SEND_SIGNAL(src, COMSIG_SUBSYSTEM_POST_INITIALIZE)
+
+	if(!failed)
+		SEND_SIGNAL(src, COMSIG_SUBSYSTEM_POST_INITIALIZE)
+	else
+		can_fire = FALSE
```

Comment:
is it really worth making failing to init standard behavior? idk

Reply:
I don't know. I just don't like how SSlua handled it by not calling ..() and implementing its own version (which fell out of sync)

Reply:
I suppose a reasonable compromise is having SSlua complete normally but also set `can_fire` to false and maybe print an extra message along with the exception. Would you prefer that, @LemonInTheDark ?

Reply:
let's wait for mso to weigh in, it's his work after all (assuming he isn't high)

Reply:
we do need a way to detect things that don't initialize correctly but given how subsystems are often used as a namespace for global vars, its entirely possible for init and fire to never touch the same variables meaning a failure in one doesn't always begate a failure in others.

I'm thinking enforced return values.

base subsystem init returns -1 (SS_INIT_NONE) (which should print a warning message like the base fire() does)

subsystems should return 2 (SS_INIT_SUCCESS) (prints a world message) or 3 (SS_INIT_NO_NEED) (prints no world message, for if the subsystem is disabled by config or some other condition wasn't met so it did not need to initialize) or -2 (SS_INIT_FAILURE).

returning `-2` should do the can_fire = 0 stuff and print a log message and set the subsystem's initialized to `FALSE`

returning null should be treated as a signal that a runtime happened in the subsystem's init proc and should print a warning message but proceed as if the init was successful.

return anything else should be treated as null with the addition of printing an extra message about invalid init return value.

either way I don't like (ab)using parent calls for signaling like that when we have an outer layer to do signalling properly (with return values).

Reply:
I can work on that.

Reply:
@MrStonedOne I've implemented what I believe is the spec you've requested. Due to touching every subsystem, I'd recommend a test merge.